pkg/kube/multicluster/secretcontroller.go

```diff
@@ -167,6 +170,13 @@ func (c *Controller) AddHandler(h ClusterHandler) {
 
 // Run starts the controller until it receives a message over stopCh
 func (c *Controller) Run(stopCh <-chan struct{}) error {
+	// wait for namespace informer synced
+	if !kube.WaitForCacheSync(stopCh, c.namespaces.HasSynced) {
+		log.Error("Failed to sync namespaces")
+		return nil
+	}
+	// run discovery namespace filter initiation before kube registry run
+	c.DiscoveryNamespacesFilter.SyncNamespaces()
```

Comment:
So we need this because we don't enable the handlers for it until initDiscoveryHandlers in kube/controller...?

That seems a bit odd. Why don't we just add the handlers here and wait for it to sync? 

I guess its maybe wrong since the queue would be different.

Overall it seems the design of DiscoveryFilter may be causing some issues. I would think we should have the DiscoveryFilter work on its own (and we run it fully here), and allow other controllers to register callbacks to it?

On a related note - do other users like NamespaceController react to changes in DiscoveryFilter properly? I suspect there is some issues there..

Reply:
Actually `initDiscoveryHandlers ` is just register namespace handler, we can not gurantee the executing order.

Reply:
>On a related note - do other users like NamespaceController react to changes in DiscoveryFilter properly? I suspect there is some issues there..

There is `handleSelectedNamespace` and `handleDeselectedNamespace` which handles that

Reply:
SyncNamespaces - Am I reading this correctly - This is never used currently and we only use FilterNamespace when namespace controller reads namespace?

Reply:
Yes, currently dead code

Reply:
> There is handleSelectedNamespace and handleDeselectedNamespace which handles that

This only handles pods, service, and endpoints. We use the discovery filter all over now - configmap, ingress, etc

Reply:
You are right, can you file an issue to track it. I think it may be a little tricky to resolve it

Reply:
IMO the current implementation of discovery filter is completely broken. I am having a hard time understanding if this PR helps or hurts, since we cannot add tests since the tests fail since the implementation is already broken...